### PR TITLE
EY/MIN-261. As a System I want to calculate Liquidation pool size based on loans in the pool

### PR DESCRIPTION
**Pull request by Jekshmek**
**Description:**

AS a System
WHEN calculating Liquidation pools ideal state
THEN I take as a basis value not the pool size, but the total borrow of the Liquidity pool

**Changes in storage or API:**

Describe storage structure changes or API changes. Make sure to update the [API docs](https://minterestfinance.atlassian.net/wiki/spaces/MINTEREST/pages/134807557/Minterest+API).

**Checklist**

- [x] Code is covered with tests
- [ ] Code is covered with benchmarks
- [ ] Fresh benchmark results included

### CODE REVIEW

File: pallets/liquidation-pools/src/lib.rs [hunk 1]
```diff
@@ -638,7 +638,7 @@ impl<T: Config> Pallet<T> {
 		let oracle_price = T::PriceSource::get_underlying_price(pool_id).ok_or(Error::<T>::InvalidFeedPrice)?;
 		let balance_ratio = Self::liquidation_pool_data_storage(pool_id).balance_ratio;
 		// Liquidation pool ideal balance in USD: liquidity_pool_balance * balance_ratio * oracle_price
```
Comment (lut1k) [inline pallets/liquidation-pools/src/lib.rs]:
```suggestion
		// Liquidation pool ideal balance in USD: liquidity_pool_borrow_usd * balance_ratio * oracle_price
```
Also fix comments for this function in line 635: `liquidity_pool_borrow_usd`

Comment (lut1k) [reply pallets/liquidation-pools/src/lib.rs]: **In** tests for liquidation-pools, you need to correct comments:
```
/*
			Liquidity Pools **borrows** (in assets): [500_000, 1_000_000, 1_500_000, 2_000_000]
			Liquidity Pools **balances** (in USD): [500_000, 2_000_000, 7_500_000, 20_000_000]
			Ideal balances 0.2 * liquidity_pool_borrow (in USD): [100_000, 400_000, 1_500_000, 4_000_000]

			Liquidation Pools balances (in assets): [400_000, 300_000, 200_000, 100_000]
			Liquidation Pools balances (in USD): [400_000, 600_000, 1_000_000, 1_000_000]

			Sales list (in assets): [(DOT, BTC, 300_000$), (KSM, BTC, 200_000$)]

			*/
```
check please all comments in test.rs

File: pallets/liquidation-pools/src/mock.rs [hunk 2]
```diff
@@ -159,6 +159,14 @@ impl Default for ExternalityBuilder {
...
 					},
 				),
+				(
+					KSM,
+					PoolData {
+						borrowed: Balance::zero(),
+						borrow_index: Rate::one(),
+						protocol_interest: Balance::zero(),
+					},
+				),
```
Comment (lut1k) [inline pallets/liquidation-pools/src/mock.rs]: Suggestion: remove it

File: pallets/liquidation-pools/src/mock.rs [hunk 3]
```diff
@@ -204,27 +212,28 @@ impl ExternalityBuilder {
...
-		self.liquidity_pools.push((
-			pool_id,
-			PoolData {
-				borrowed: Balance::zero(),
-				borrow_index: Rate::one(),
-				protocol_interest: Balance::zero(),
-			},
-		));
+	pub fn pool_remove(mut self, pool_id: CurrencyId) -> Self {
+		self.liquidity_pools.retain(|&(currency_id, _)| currency_id != pool_id);
```
Comment (lut1k) [inline pallets/liquidation-pools/src/mock.rs]: Suggestion: remove it

File: pallets/liquidation-pools/src/mock.rs [hunk 4]
```diff
@@ -204,27 +212,28 @@ impl ExternalityBuilder {
...
+				*pool = PoolData {
+					borrowed: balance,
+					borrow_index: Rate::one(),
+					protocol_interest: Balance::zero(),
+				};
+				return self;
+			}
+		}
 		self
 	}
```
Comment (lut1k) [inline pallets/liquidation-pools/src/mock.rs]:
```suggestion
	pub fn set_pool_borrow_underlying(mut self, currency_id: CurrencyId, balance: Balance) -> Self {
		self.liquidity_pools.push((
			currency_id,
			PoolData {
				borrowed: balance,
				borrow_index: Rate::one(),
				protocol_interest: Balance::zero(),
			},
		));
		self
	}
```

File: pallets/liquidation-pools/src/tests.rs [hunk 5]
```diff
@@ -55,7 +55,7 @@ fn offchain_worker_balancing_test() {
 
 #[test]
 fn protocol_operations_not_working_for_nonexisting_pool() {
-	ExternalityBuilder::default().build().execute_with(|| {
+	ExternalityBuilder::default().pool_remove(KSM).build().execute_with(|| {
```
Comment (lut1k) [inline pallets/liquidation-pools/src/tests.rs]: Suggestion: remove it

**Comment by lut1k** [comment]
Also, before calling the `LiquidityPoolsManager::get_pool_borrow_underlying()` function, it is necessary to call the `ControllerManager::accrue_interest_rate()` function. I think you can add call `T::ControllerManager::accrue_interest_rate(pool_id)?;` in line 495 in function `collects_sales_list()`. 
To do this, you will need to add the functionality ControllerManager ti the liquidation-pools pallet.

File: pallets/liquidation-pools/src/lib.rs [hunk 6]
```diff
@@ -637,8 +642,9 @@ impl<T: Config> Pallet<T> {
 	fn calculate_pool_ideal_balance_usd(pool_id: CurrencyId) -> BalanceResult {
 		let oracle_price = T::PriceSource::get_underlying_price(pool_id).ok_or(Error::<T>::InvalidFeedPrice)?;
 		let balance_ratio = Self::liquidation_pool_data_storage(pool_id).balance_ratio;
-		// Liquidation pool ideal balance in USD: liquidity_pool_balance * balance_ratio * oracle_price
-		let ideal_balance_usd = Rate::from_inner(T::LiquidityPoolsManager::get_pool_available_liquidity(pool_id))
+		T::ControllerManager::accrue_interest_rate(pool_id)?;
```
Comment (lut1k) [inline pallets/liquidation-pools/src/lib.rs]: Two calls `fn accrue_interest_rate()` are redundant. In my opinion you can only leave call in `fn collects_sales_list()`.

**Review by lut1k — DISMISSED** [review]

File: pallets/liquidation-pools/src/lib.rs [hunk 7]
```diff
@@ -661,8 +666,8 @@ impl<T: Config> Pallet<T> {
 	fn calculate_pool_ideal_balance_usd(pool_id: CurrencyId) -> BalanceResult {
```
Comment (kurotych) [inline pallets/liquidation-pools/src/lib.rs:712]: Should we update comment to this function? 
/// Returns minimum of (**liquidity_pool_balance** * balance_ratio * oracle_price) and
/// max_ideal_balance_usd

Comment (Jekshmek) [reply pallets/liquidation-pools/src/lib.rs:666]: @kurotych The comment does not reflect either the new or the old calculation method. I think that there is no need to update, but what do you think?

Comment (Jekshmek) [reply pallets/liquidation-pools/src/lib.rs:712]: @kurotych 
/// Returns minimum of (**liquidity_pool_total_borrow** * balance_ratio * oracle_price) and
/// max_ideal_balance_usd

**Review by lut1k — DISMISSED** [review]

File: pallets/liquidation-pools/src/tests.rs [hunk 8]
```diff
@@ -53,52 +53,6 @@ fn offchain_worker_balancing_test() {
 	});
 }
 
-#[test]
-fn offchain_worker_balancing_off_test() {
```
Comment (Apollo50) [inline pallets/liquidation-pools/src/tests.rs]: Why did you remove this test?

Comment (Jekshmek) [reply pallets/liquidation-pools/src/tests.rs]: @Apollo50 I'll restore it right now, I don't know how it happened

File: pallets/liquidation-pools/src/lib.rs [hunk 9]
```diff
@@ -687,13 +692,13 @@ impl<T: Config> Pallet<T> {
 	/// Calculates ideal balance for pool balancing
 	/// - `pool_id`: PoolID for which the ideal balance is calculated.
 	///
-	/// Returns minimum of (liquidity_pool_balance * balance_ratio * oracle_price) and
+	/// Returns minimum of (liquidity_pool_total_borrow * balance_ratio * oracle_price) and
 	/// max_ideal_balance_usd
```
Comment (lut1k) [inline pallets/liquidation-pools/src/lib.rs]:
```suggestion
	/// Returns minimum of (liquidity_pool_borrow_underlying * balance_ratio * oracle_price) and
	/// max_ideal_balance_usd
```

File: pallets/liquidation-pools/src/lib.rs [hunk 10]
```diff
@@ -687,13 +692,13 @@ impl<T: Config> Pallet<T> {
...
 	fn calculate_pool_ideal_balance_usd(pool_id: CurrencyId) -> BalanceResult {
 		let oracle_price = T::PriceSource::get_underlying_price(pool_id).ok_or(Error::<T>::InvalidFeedPrice)?;
 		let balance_ratio = Self::liquidation_pool_data_storage(pool_id).balance_ratio;
-		// Liquidation pool ideal balance in USD: liquidity_pool_balance * balance_ratio * oracle_price
-		let ideal_balance_usd = Rate::from_inner(T::LiquidityPoolsManager::get_pool_available_liquidity(pool_id))
+		// Liquidation pool ideal balance in USD: liquidity_pool_total_borrow * balance_ratio * oracle_price
```
Comment (lut1k) [inline pallets/liquidation-pools/src/lib.rs]:
```suggestion
		// Liquidation pool ideal balance in USD: liquidity_pool_borrow_underlying * balance_ratio * oracle_price
```

**Review by lut1k — APPROVED** [review]

**Review by Apollo50 — APPROVED** [review]
